jena-arq/src/main/java/org/apache/jena/sparql/engine/main/iterator/QueryIterUnion.java

```diff
@@ -51,21 +50,51 @@ public QueryIterUnion(QueryIterator input,
     @Override
     protected QueryIterator nextStage(Binding binding)
     {
-        QueryIterConcat unionQIter = new QueryIterConcat(getExecContext()) ;
-        for (Op subOp : subOps)
-        {
-            subOp = QC.substitute(subOp, binding) ;
-            QueryIterator parent = QueryIterSingleton.create(binding, getExecContext()) ;
-            QueryIterator qIter = QC.execute(subOp, parent, getExecContext()) ;
-            unionQIter.add(qIter) ;
-        }
-        
-        return unionQIter ;
+        Iterator<Op> subOpIt = subOps.iterator();
+        return new QueryIter(getExecContext()) {
+            QueryIterator qIter = null;
+
+            @Override
+            protected void requestCancel() {
+                performRequestCancel(qIter);
+            }
+
+            @Override
+            protected Binding moveToNextBinding() {
+                return qIter.next();
+            }
+
+            @Override
+            protected boolean hasNextBinding() {
+                for (;;) {
+                    if (qIter != null) {
+                        if (qIter.hasNext()) {
+                            return true;
+                        } else {
+                            qIter.close();
+                            qIter = null;
+                        }
+                    } else {
+                        if (subOpIt.hasNext()) {
+                            Op subOp = subOpIt.next();
+                            qIter = QC.execute(subOp, binding, getExecContext());
+                        } else {
+                            return false;
+                        }
+                    }
+                }
+            }
```

Comment:
Making the nextStage() method of `QueryIterUnion` truly lazy removes the need for try-catch blocks: Sub-iterators are only opened when needed, and if a creation fails then there are no previously opened resources that need to be cleaned up.